tuner/tuner/spec_builder.py

```diff
@@ -24,12 +38,43 @@ def build_td_spec(
     func_name: str,
 ) -> ir.Module:
     bbargs = []
+    # The `root_op` attribute will prevent matching of ops without the attr in
+    # the resulting TD spec matcher if it is not removed, so we remove it here.
+    # After removing, we must add it back, since the op is connected to the
+    # input module, which gets used for all candidates.
+    # TODO(Max191): Find a cleaner way to do this without removing and adding
+    # back the attribute.
+    has_root_attr = ROOT_OP_ATTR_NAME in op.opview.attributes
+    if has_root_attr:
+        assert isinstance(
+            op.opview.attributes[ROOT_OP_ATTR_NAME], ir.UnitAttr
+        ), f"expected '{ROOT_OP_ATTR_NAME}' attr to be a unit attr"
+    if has_root_attr:
+        del op.opview.attributes[ROOT_OP_ATTR_NAME]
+    # Get the root op string for formatting the final spec.
+    root_operation = str(op)
+    if has_root_attr:
+        op.opview.attributes[ROOT_OP_ATTR_NAME] = ir.UnitAttr.get(op.context)
+
+    # Get the names ssa names of operands to make sure they match in the
+    # template after string formatting.
+    captured_values: set[ir.Value] = set()
     for operand in op.operands:
+        if operand in captured_values:
+            # TODO(Max191): Remove this warning when the transform for the
+            # `cast_compatible_dag_from_root` op fixes a bug in the matching
+            # logic that causes failure to match when the same operand is
+            # repeated. For now, still avoid adding duplicate SSA values to
+            # prevent parsing failure.
```

Comment:
There is a bug in the transform op, which I think is a very simple fix, but we need to land it in IREE before this works properly for repeated operands.